src/lib/phdi-building-blocks/phdi_building_blocks/linkage.py

```diff
@@ -1,68 +1,100 @@
 import hashlib
+import copy
+from phdi_building_blocks.utils import find_patient_resources, get_one_line_address
 
 
-def add_patient_identifier(bundle: dict, salt_str: str) -> dict:
+def add_linking_identifier_to_patients_in_bundle(
+    bundle: dict, salt_str: str, overwrite: bool = True
+) -> dict:
     """
-    Given a FHIR resource bundle (defined as a dictionary of resources
-    containing at least one patient resource):
-      - identify the patient resource(s) in the bundle
+    Given a FHIR resource bundle:
+      - identify all patient resource(s) in the bundle
       - extract standardized name, DOB, and address information for each
       - compute a unique hash string based on these fields
       - add the hash string to the list of identifiers held in that
         patient resource
-    This function assumes data has been standardized already by the
-    silver transforms.
+    :param dict bundle: The FHIR bundle for whose patients to add a
+    linking identifier
+    :param str salt_str: The suffix string added to prevent being
+    able to reverse the hash into PII
+    :param bool overwrite: Whether to write the new standardizations
+    directly into the given bundle, changing the original data (True
+    is yes)
+    :return: The bundle of data with patients having unique identifiers
+    :rtype: dict
     """
+    # Copy the data if we don't want to overwrite the original
+    if not overwrite:
+        bundle = copy.deepcopy(bundle)
 
-    for resource in bundle["entry"]:
-        if resource["resource"]["resourceType"] == "Patient":
-            patient = resource["resource"]
+    for resource in find_patient_resources(bundle):
+        patient = resource.get("resource")
 
-            # Combine given and family name
-            recent_name = next(
-                (name for name in patient["name"] if name.get("use") == "official"),
-                patient["name"][0],
-            )
+        # Combine given and family name
+        recent_name = get_field_with_use(patient, "name", "official", 0)
+        name_parts = recent_name.get("given", []) + [recent_name.get("family")]
```

Comment:
What if the `family` name isn't provided? Would this result in `name_parts = [] + [None]`? Is the `given` name _always_ in a list?

Reply:
From what we've seen, yeah given name is always in a list (not only does given include middle name, but it's list structured for people with hyphenated or multiple first names). I've added a default empty string to the `get` for family name, so that if no name is provided we wind up with `[""]` as `name_parts`. When we later call the `join` method to build the name string, this will lead to the final value being an empty string, as is desired.